bootloader/grub.go

```diff
@@ -379,11 +386,103 @@ func (g *grub) ManagedAssets() []string {
 	}
 }
 
-// CommandLine returns the kernel command line composed of the built-in
-// list and extra arguments passed in arguments. The command line may be
-// different when using a bootloader in the recovery partition.
+// CommandLine returns the kernel command line composed of system run mode
+// arguments, extra arguments (configured by snapd) and the built-in bootloader
+// specific arguments. The command line may be different when using a bootloader
+// in the recovery partition.
 //
 // Implements ManagedAssetsBootloader for the grub bootloader.
-func (g *grub) CommandLine(extra []string) (string, error) {
-	return "", fmt.Errorf("not implemented")
+func (g *grub) CommandLine(modeArgs, extraArgs string) (string, error) {
+	// we do not trust the on disk asset, use the built-in one
+	assetName := "grub.cfg"
+	isRecovery := false
+	if g.recovery {
+		assetName = "grub-recovery.cfg"
+		isRecovery = true
+	}
+	staticCmdline, err := staticCommandLineFromGrubAsset(assetName)
+	if err != nil {
+		return "", fmt.Errorf("cannot extract static command line element: %v", err)
+	}
+	// sort arguments so that they match their positions
+	mode, err := modeArgsForGrub(modeArgs, isRecovery)
+	if err != nil {
+		return "", err
+	}
+	args, err := strutil.KernelCommandLineSplit(staticCmdline + " " + extraArgs)
+	if err != nil {
+		return "", fmt.Errorf("cannot use badly formatted kernel command line: %v", err)
+	}
+	// join all argument with a single space, see
+	// grub-core/lib/cmdline.c:grub_create_loader_cmdline() for reference,
+	// arguments are separated by a single space, the space after last is
+	// replaced with terminating NULL
+	return strings.Join(append(mode, args...), " "), nil
+}
+
+// static command line is defined as:
+//     set snapd_static_cmdline_args='arg arg arg'\n
+// or
+//     set snapd_static_cmdline_args='arg'\n
+const grubStaticCmdlinePrefix = `set snapd_static_cmdline_args=`
+const grubStaticCmdlineQuote = `'`
+
+// staticCommandLineFromGrubAsset extracts the static command line element from
+// grub boot config asset on disk.
+func staticCommandLineFromGrubAsset(asset string) (string, error) {
+	gbc := assets.Internal(asset)
+	if gbc == nil {
+		return "", fmt.Errorf("internal error: asset %q not found", asset)
+	}
+	scanner := bufio.NewScanner(bytes.NewReader(gbc))
```

Comment:
I wonder if a good old fashioned regex would be easier to read here, something like:
```
var grubStaticCmdlineArgsRE = regexp.MustCompile(`(?ms)^set snapd_static_cmdline_args='(.*)'$`)

// staticCommandLineFromGrubAsset extracts the static command line element from
// grub boot config asset on disk.
func staticCommandLineFromGrubAsset(asset string) (string, error) {
	gbc := assets.Internal(asset)
	if gbc == nil {
		return "", fmt.Errorf("internal error: asset %q not found", asset)
	}
	m := grubStaticCmdlineArgsRE.FindSubmatch(gbc)
	switch len(m) {
	case 0:
		return "", nil
	case 2:
		return string(m[1]), nil
	default:
		return "", fmt.Errorf("found multiple snapd_static_cmdline_args declarations")
	}
}
```
This lacks the extra error checking you added so maybe with that it's not that much of a difference.
But it seems easier to read then the index juggling :)

Reply:
Nice, looks way simpler.

Reply:
Applied with some tweaks, dopped `(s)` otherwise it matched too much and added some `'<quote>'` checks around.